server/src/main/java/org/opensearch/index/query/QueryBuilderVisitor.java

```diff
@@ -18,8 +18,9 @@ public interface QueryBuilderVisitor {
     /**
      * Accept method is called when the visitor accepts the queryBuilder object to be traversed in the query tree.
      * @param qb is a queryBuilder object which is accepeted by the visitor.
+     * @param level level of the current query builder based on the QueryBuilder tree. 0 is top/root level.
      */
-    void accept(QueryBuilder qb);
+    void accept(QueryBuilder qb, int level);
```

Comment:
We don't need to change the signature of `accept` to pass `level`. We can track levels in the QueryVisitor itself.

In your SearchQueryCategorizer, you should be able to do something like:

```
class SearchQueryCategorizingVisitor implements QueryBuilderVisitor {
  private final int level;
  private final SearchQueryCounters counters;

  public SearchQueryCategorizingVisitor(SearchQueryCounters counters) {
    this(counters, 0);
  }

  private SearchQueryCategorizingVisitor(SearchQueryCounters counters, int level) {
    this.counters = counters;
    this.level = level;
  }
  public void accept(QueryBuilder qb) {
    // Do the whole implementation that you defined above, using the level member
  }

  public QueryBuilderVisitor getChildVisitor(BooleanClause.Occur occur) {
    return new SearchQueryCategorizingVisitor(counters, level + 1);
  }
}
```

Reply:
Thanks @msfroh . Let me try this out